Machine Translation module preview

Screenfeed opened the pull request:
This PR adds a "Machine Translation" preview module.
This also brings changes in `PLL_Settings_Module` and its property `$active_option`, to prevent duplicating code in modules.

The modification in CSS is for the `vertical-align` to be reused in the news settings fields in PLL Pro.

Chouby commented:
![image](https://github.com/polylang/polylang/assets/3238583/87548aa5-7b85-4009-932c-fda20ee17a2e)
I expected that the module appears as deactivated as the others.

manooweb commented on js/src/admin.js:
```diff
@@ -2,6 +2,67 @@
+				return;
+			}
+
+			const sep = ( ajaxurl.indexOf( '?' ) !== -1 ) ? '&' : '?';
```
See if we couldn't also depend on `wp.url` package? 🤔
There are also functions available in this package like `addQueryArgs` which already manage the query string separator.

Screenfeed replied:
Done in [6ceaf75](https://github.com/polylang/polylang/pull/1420/commits/6ceaf753e97aba3322e52be876406ff58e3d5424).

Screenfeed commented:
> I expected that the module appears as deactivated as the others.

Fixed in [369842c](https://github.com/polylang/polylang/pull/1420/commits/369842c57df04ac79cb26b69b6c1c2b908f9604b).
I introduced some improvements to `PLL_Settings_Module`, to prevent duplicating code in each "preview" class.
For example this avoids to overwrite `is_active()` in `PLL_Settings_Preview_Machine_Translation` (like we did in `PLL_Settings_Preview_Share_Slug`) to make it always return `false`, and then re-overwrite it again in PLL Pro to re-enable the original method.

Screenfeed commented on js/src/admin.js:
```diff
@@ -33,16 +33,16 @@ window.pll.settings_ajax_button = () => {
-			const sep = ( ajaxurl.indexOf( '?' ) !== -1 ) ? '&' : '?';
-			let   url = ajaxurl + sep + 'action=' + action + '&_pll_nonce=' + nonce + '&pll_ajax_settings=1';
-
+			let urlParams = { 'action': action, '_pll_nonce': nonce, 'pll_ajax_settings': 1 };
```
Fixed in [3314243](https://github.com/polylang/polylang/pull/1420/commits/33142432e8760b53abf6b1ea3b09b0b2efa684f2).

Hug0-Drelon commented on js/src/admin.js:
```diff
@@ -2,6 +2,67 @@
  * @package Polylang
  */
 
+window.pll = window.pll || {};
```
This is an old fashion way of accessing `pll` from other script sheets. This is not what we want here.
Also it would break without error if some third party do the same.
Please use `const pllMachineTranslationButton` or something instead.

Screenfeed replied:
It wouldn't break unless the plugin does this directly:
```JS
window.pll = { ... };
```
In this case we can say it's intentional.

`const` added in [411277a](https://github.com/polylang/polylang/pull/1420/commits/411277a5b02769446f93f41ed8cbed2522dddc41).

Hug0-Drelon commented on js/src/admin.js:
```diff
@@ -2,6 +2,67 @@
+/**
+ * Settings: manages AJAX calls attached to `.pll-ajax-button` buttons.
+ */
+window.pll.settings_ajax_button = () => {
```
I see you have a modular approach so I suggest to go all in.
- Move the code in a dedicated file then import it here.
- Have a `init()` function so you elements are prepared before the event listener callback is called.

Bonus, camelCase please ❤️

Screenfeed replied:
Done in [411277a](https://github.com/polylang/polylang/pull/1420/commits/411277a5b02769446f93f41ed8cbed2522dddc41) and [9ce6f69](https://github.com/polylang/polylang/pull/1420/commits/9ce6f69003c2b3d8c8f7a5f4d9a3d8a881be5249).

Hug0-Drelon commented on js/src/admin.js:
```diff
@@ -2,6 +2,67 @@
+				}
+			} ).catch( () => {
+				button.removeAttribute( 'disabled' );
+				errorElms[0].textContent = 'Unknown error.';
```
Maybe we want to translate that.

Screenfeed replied:
Done in [411277a](https://github.com/polylang/polylang/pull/1420/commits/411277a5b02769446f93f41ed8cbed2522dddc41).

Hug0-Drelon commented on js/src/admin.js:
```diff
@@ -2,6 +2,67 @@
+				return response.json();
+			} ).then( ( json ) => {
+				if ( json.success ) {
+					row.classList.add( 'notice-success', 'notice-alt' );
```
This kind of logic could benefit from a `toggle` function along `button.removeAttribute( 'disabled' );`.

Screenfeed replied:
I didn't went through, it adds code instead of removing some.

manooweb replied:
🤔 we can use `toggleAttribute( 'disabled')` everywhere instead of `setAttribute` or `removeAttribute`
See https://developer.mozilla.org/en-US/docs/Web/API/Element/toggleAttribute

Screenfeed replied:
Nope, I'm against any `toggle*()`, it's too unpredictable: when you use it don't know if it will add or remove the attribute.

Hug0-Drelon commented on js/src/admin.js:
```diff
@@ -2,61 +2,7 @@
-};
-
-pllMachineTranslationAjaxButton.init();
+import {} from './settings';
```
```suggestion
import { pllMachineTranslationAjaxButton } from './settings';

pllMachineTranslationAjaxButton.init();
```
So it describe what it does, and run the code.

manooweb replied:
Please put `settings.js` in lib sub folder not to be built itself.

Screenfeed replied:
> Please put `settings.js` in `lib` sub folder not to be built itself.

Done in [f916e62](https://github.com/polylang/polylang/pull/1420/commits/f916e6260bd5b55597307fa82b2d20bb6d66038a).

manooweb commented on modules/machine-translation/settings-preview-machine-translation.php:
```diff
@@ -0,0 +1,39 @@
+		parent::__construct(
+			$polylang,
+			array(
+				'module'        => 'machine-translation',
```
Settings actions are based on module name.
For example https://github.com/polylang/polylang/blob/3.5.4/settings/settings.php#L260-L266
When Polylang Pro is activated the module name is the full namespaced qualified class name
https://github.com/polylang/polylang-pro/blob/ef45ad2f04073a66397fa52011d3bd78e87aa85f/modules/machine-translation/load.php#L27-L31 
This is this name which is used as key in the settings `modules` array because the class name doesn't starts with `PLL_Settings_` 
See https://github.com/polylang/polylang/blob/3.5.4/settings/settings.php#L89
It is also used as module property in the instance itself 
https://github.com/polylang/polylang/blob/3.5.4/settings/settings-module.php#L116

So sanitization make this module name unknown 
https://github.com/polylang/polylang/blob/3.5.4/settings/settings.php#L262 because `wp_syntex\\polylang_pro\\modules\\machine_translation\\module_settings` is sanitized to `wp_syntexpolylang_promodulesmachine_translationmodule_settings`

Screenfeed replied:
Fixed in [ed560d7](https://github.com/polylang/polylang/pull/1420/commits/ed560d77b66a28f77553259469920d5aaeb16079).
@manooweb You were not looking at the right git branch, this system [has changed a bit](https://github.com/polylang/polylang-pro/blob/f5251bbd3e47ed4d396c8475af4c52cc045a8cf4/modules/admin-loader/admin-loader.php#L91-L92).

Screenfeed replied:
@manooweb Ah! You're right, I missed it.

Screenfeed replied:
Fixed in [c284291](https://github.com/polylang/polylang-pro/pull/1974/commits/c284291b6ac6e992b7b36c86aa0a807bc4ab2924).

Chouby commented on modules/machine-translation/settings-preview-machine-translation.php:
```diff
@@ -0,0 +1,39 @@
+	 *
+	 * @var int
+	 */
+	public $priority = 85;
```
I think it's free and still < License key prio:
```suggestion
	public $priority = 90;
```

Screenfeed replied:
Done in [72fb3a5](https://github.com/polylang/polylang/pull/1420/commits/72fb3a59c4b05435a120e5467ff525323cf5fe15).

Chouby commented on modules/machine-translation/settings-preview-machine-translation.php:
```diff
@@ -0,0 +1,39 @@
+	 *
+	 * @param PLL_Settings $polylang Polylang object.
+	 */
+	public function __construct( &$polylang ) {
+		parent::__construct(
+			$polylang,
+			array(
+				'module'        => 'machine_translation',
+				'title'         => __( 'Machine Translation', 'polylang' ),
+				'description'   => __( 'Allows linkage to an external translation solution.', 'polylang' ),
+				'active_option' => '',
+			)
+		);
+	}
```
For previews like this one (for share slugs and translated slugs too), I suggest to merge current arguments with arguments coming from the child class:
```suggestion
	public function __construct( &$polylang, $args = array() ) {
		$defaults = array(
			'module'        => 'machine_translation',
			'title'         => __( 'Machine Translation', 'polylang' ),
			'description'   => __( 'Allows linkage to an external translation solution.', 'polylang' ),
			'active_option' => '',
		);

		parent::__construct( $polylang, array_merge( $default, $args ) );
	}
```
What do you think

Screenfeed replied:
Done in [9732372](https://github.com/polylang/polylang/pull/1420/commits/97323722ce8c9f85ede58ff4d9bdc3bc31584b72).

Screenfeed commented:
Following the discussion with @Chouby, `PLL_Settings_Module::$active_option` can now have 3 values: `'none'` (is always active), `'preview'` (the module is a preview in PLL, so, always, inactive), and an option key (can be activated/deactivated). In PLL Pro, a `'preview'` module must be overwritten by one of the other 2 values.

Chouby commented on settings/settings-module.php:
```diff
@@ -145,62 +170,62 @@ public function __construct( &$polylang, $args ) {
 	 */
 	public function is_active() {
-		return empty( $this->active_option ) || ! empty( $this->options[ $this->active_option ] );
+		return 'none' === $this->active_option || 'preview' !== $this->active_option && ! empty( $this->options[ $this->active_option ] );
```
Could you add parentheses? I don't like relying on operator precedence.

Screenfeed replied:
Done in [daeb7eb](https://github.com/polylang/polylang/pull/1420/commits/daeb7eb86853195c76675b76bd5141c96be20faa).

Chouby commented on settings/settings-module.php:
```diff
@@ -29,9 +29,15 @@ class PLL_Settings_Module {
+	 * Key to use to manage the module activation state.
+	 * Possible values:
+	 * - An option key for a module that can be activated/deactivated.
+	 * - 'none' for a module that is always active.
```
I know that I wrote that some days ago but it seems that's incorrect. Instead it looks like 'none' corresponds to "no settings available". Whether it is active or not is programmatically decided. See `PLL_Settings_Browser` for ex.

Screenfeed replied:
You're right, I forgot this case. Done in [01e6886](https://github.com/polylang/polylang/pull/1420/commits/01e6886688f096099ab78716d813791289a8daf1).
